TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/logic/backfill.py

```diff
@@ -0,0 +1,126 @@
+from enum import Enum
+import json
+import logging
+from typing import Dict, Optional
+
+from console_link.models.backfill_osi import OpenSearchIngestionBackfill
+from console_link.models.backfill_rfs import DockerRFSBackfill, ECSRFSBackfill, RFSBackfill
+from console_link.models.cluster import Cluster
+from console_link.models.backfill_base import Backfill
+import yaml
+
+
+logger = logging.getLogger(__name__)
+
+BackfillType = Enum("BackfillType",
+                    ["opensearch_ingestion", "reindex_from_snapshot"])
+
+
+class UnsupportedBackfillTypeError(Exception):
+    def __init__(self, supplied_backfill: str):
+        super().__init__("Unsupported backfill type", supplied_backfill)
+
+
+def get_backfill(config: Dict, source_cluster: Optional[Cluster], target_cluster: Optional[Cluster]) -> Backfill:
+    if BackfillType.opensearch_ingestion.name in config:
+        if source_cluster is None:
+            raise ValueError("source_cluster must be provided for OpenSearch Ingestion backfill")
+        if target_cluster is None:
+            raise ValueError("target_cluster must be provided for OpenSearch Ingestion backfill")
+        logger.debug("Creating OpenSearch Ingestion backfill instance")
+        return OpenSearchIngestionBackfill(config=config,
+                                           source_cluster=source_cluster,
+                                           target_cluster=target_cluster)
+    elif BackfillType.reindex_from_snapshot.name in config:
+        if target_cluster is None:
+            raise ValueError("target_cluster must be provided for RFS backfill")
+
+        if 'docker' in config[BackfillType.reindex_from_snapshot.name]:
+            logger.debug("Creating Docker RFS backfill instance")
+            return DockerRFSBackfill(config=config,
+                                     target_cluster=target_cluster)
+        elif 'ecs' in config[BackfillType.reindex_from_snapshot.name]:
+            logger.debug("Creating ECS RFS backfill instance")
+            return ECSRFSBackfill(config=config,
+                                  target_cluster=target_cluster)
+
+    logger.error(f"An unsupported metrics source type was provided: {config.keys()}")
+    if len(config.keys()) > 1:
+        raise UnsupportedBackfillTypeError(', '.join(config.keys()))
+    raise UnsupportedBackfillTypeError(next(iter(config.keys())))
+
+
+def describe(backfill: Backfill, as_json=False) -> str:
```

Comment:
I'm curious about this paradigm - why wrap the Backfill in this way?  Having the additional layer here could be surprising to folks and I'm interested to hear what, if anything, would go wrong if someone came in and used the object directly?  

As an alternative, why not have the return-type modularity exist in something like an OutputType enum or an Outputter object that the Backfill object contains?

Reply:
Basically, why the logic layer?

The idea is two-fold:
1. The CLI and API are intended to be very thin clients, potentially generated automatically from a spec. This logic/middleware layer will be the layer that those clients interact with, so all they have to know is what functions are available from the library. Potentially, these files can even be used as the spec to generate the clients.
2. The CLI and API should never have to know exactly what underlying backend adapters they're dealing with. But in some cases (e.g. CLI params that only apply to one type) _someone_ in between the CLI and the class itself needs to know what's happening under the hood. That's what can happen in this layer.

Someone who's building on top of the library directly is welcome to use the underlying classes, but then they're taking on responsibility for that.

Very open to feedback on a better way to do it.

For the return type specifically, I like the idea of standardizing it more--I'm not sure I quite understand what you mean by "an Outputter object that the Backfill object contains"